Integrate Fermionic QAOA(FQAOA) Module and Associated Files into OpenQAOA

yoshioka1128 opened the pull request:
## Description

This pull request introduces the FQAOA (Fermionic Quantum Approximate Optimization Algorithm) module along with its associated files into the OpenQAOA framework. The integration includes the following components:
- fqaoa_workflow.py: Implements the FQAOA workflow, enabling the use of fermionic methods within the OpenQAOA framework.
- fqaoa_utils.py: Provides utility functions to support the FQAOA implementation, including essential operations and helper functions.
- test_fqaoa.py: Contains unit tests for the fqaoa_utils.py, ensuring the correctness and reliability of the new functionality.
 test_workflows.py: Tests the overall workflow integration, confirming that FQAOA integrates smoothly with existing components.
- test_analytical_simulator.py: Includes tests for the analytical simulator, verifying the exception handling in the FQAOA approach.
- 16_FQAOA_examples.ipynb: A Jupyter notebook demonstrating an example of solving the portfolio optimization problem using FQAOA. This notebook showcases the practical application of the FQAOA method and provides a step-by-step guide for users.

All tests have been successfully executed, demonstrating the stability and effectiveness of the integrated features. This addition enhances the OpenQAOA framework by introducing the capability to handle constrained combinatorial optimization problems using FQAOA.

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [x] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [x] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

The following unit tests have been added to ensure the robustness and accuracy of the new features:
test_fqaoa.py: Tests the core functionality of the fqaoa_utils.py.
 test_workflows.py: Validates the integration of fqaoa_workflow.py within the overall workflow. 
test_analytical_simulator.py: Confirms the exception handling in the FQAOA approach.

yoshioka1128 committed:
Add __init__.py files

yoshioka1128 committed:
Add fqaoa_workflow.py

yoshioka1128 committed:
Add fqaoa_utils.py

yoshioka1128 committed:
Add test_fqaoa.py and ensure all tests pass

yoshioka1128 committed:
Add test_analytical_simulator.py and ensure all tests pass

yoshioka1128 committed:
Add test_workflows.py and ensure all tests pass

yoshioka1128 committed:
new file:   16_FQAOA_examples.ipynb

yoshioka1128 committed:
Add missing imports

yoshioka1128 committed:
Update documentation for FQAOA

yoshioka1128 committed:
new file:   17_demonstration_of_quantum_annealing_with_FQAOA.ipynb

yoshioka1128 committed:
Rename file: 17_demonstration_of_quantum_annealing_with_FQAOA.ipynb

yoshioka1128 commented:
I've just pushed an additional notebook `17_FQAOA_advanced_parameterization.ipynb` to this branch that provides a tutorial on advanced circuit parametrization, including Fourier parametrization for FQAOA. This notebook is intended to demonstrate more complex usage scenarios of FQAOA and complement the existing documentation.

KilianPoirier commented on docs/source/index.rst:
```diff
@@ -171,6 +171,18 @@ Your first RQAOA workflow
+   fqaoa.fermi_compile(qubo_problem, n_fermions)
+   fqaoa.optimize()
+
+FQAOA intrinsically imposes a hard constraint where the hamming wieght is equal to n_fermions.
```
Typo on "weight"

yoshioka1128 replied:
Fix this typo.
https://github.com/tech-sketch/openqaoa/commit/6d2b9b5301ed24ca79549575848fdab873ccf429#diff-4eec6cec5f6fab1548b85433ea8ca81315ae165db4b7f84019f287df9015699fL184-R184

KilianPoirier commented on docs/source/index.rst:
```diff
@@ -305,6 +317,7 @@ Contents
    notebooks/14_qaoa_benchmark.ipynb
    notebooks/X_dumping_data.ipynb
    notebooks/15_Zero_Noise_Extrapolation.ipynb
+   notebooks/16_FQAOA_examples.ipynb
 
```
You probably need to add `notebooks/17_advanced_parametrization.ipynb` too. The sphinx compiler is throwing a warning for this.

yoshioka1128 replied:
Add the 17_FQAOA_advanced_parameterization.ipynb.
https://github.com/tech-sketch/openqaoa/commit/6d2b9b5301ed24ca79549575848fdab873ccf429#diff-4eec6cec5f6fab1548b85433ea8ca81315ae165db4b7f84019f287df9015699fL320-R321

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
```diff
@@ -0,0 +1,764 @@
+    "\n",
+    "## A brief introduction\n",
+    "\n",
+    "We present an implementation of a novel algorithm designed for solving combinatorial optimization problems with constraints, utilizing the principles of quantum computing. The algorithm, known as the FQAOA [1, 2], introduces a significant enhancement over traditional methods by leveraging fermion particle number preservation. This intrinsic property allows the algorithm to enforce constraint ...
```
Can you link [1] and [2] to the papers directly using markdown syntax?

yoshioka1128 replied:
Link  [1] and [2] to the papers directly.
https://github.com/tech-sketch/openqaoa/blob/yoshioka1128/dev_clean_fqaoa/examples/16_FQAOA_example.ipynb?short_path=31c6b79#L15-L16

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
```diff
@@ -0,0 +1,764 @@
+    "\n",
+    "### One-Dimensional Mixer Hamiltonian\n",
+    "\n",
+    "The FQAOA implemented on OpenQAOA employs mixer Hamiltonians on one-dimensional lattices. The main features of this computational model are summarized based on the study [2]:\n",
```
Same here, add link for [2]

yoshioka1128 replied:
Add link for [2].
https://github.com/tech-sketch/openqaoa/blob/yoshioka1128/dev_clean_fqaoa/examples/16_FQAOA_example.ipynb?short_path=31c6b79#L65

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
```diff
@@ -0,0 +1,764 @@
+    "\n",
+    "- Utilises a mixer Hamiltonian on a one-dimensional lattice dedicated to combinatorial optimization problems with equality constraints of the same coefficients.\n",
+    "\n",
+    "- Reduced Gate Operations: The new mixer Hamiltonian significantly reduces the number of gate operations in quantum circuits compared to previous studies [1], thus improving computational efficiency.\n",
```
Link to [1]

yoshioka1128 replied:
Add link to [1].
https://github.com/tech-sketch/openqaoa/blob/yoshioka1128/dev_clean_fqaoa/examples/16_FQAOA_example.ipynb?short_path=31c6b79#L69

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
```diff
@@ -0,0 +1,764 @@
+    "$$\\hat{\\cal H}_M=\\sum_{i=1}^{N}\\varepsilon_i\\hat{\\gamma}_i^\\dagger\\hat{\\gamma}_i \\qquad {\\rm with} \\quad\n",
+    "\\hat{\\gamma}_i^\\dagger = \\sum_{j=1}^N[\\phi_0]_{i,j}\\hat{c}^\\dagger_j,$$\n",
+    "where $[\\phi_0]$ is the unitary matrix used for the diagonalization and $\\varepsilon_i$ is eigenvalue.\n",
+    "The formulation and validation of the model on cyclic lattice are detailed in the study [2].\n",
```
Link for [2]

yoshioka1128 replied:
Add link for [2].
https://github.com/tech-sketch/openqaoa/blob/yoshioka1128/dev_clean_fqaoa/examples/16_FQAOA_example.ipynb?short_path=31c6b79#L84

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
```diff
@@ -0,0 +1,764 @@
+    "\\prod_{i\\ {\\rm odd}}\\exp\\left[-i\\beta t\\left(\\hat{c}^{\\dagger}_i\\hat{c}_{i+1}+\\hat{c}^{\\dagger}_{i+1}\\hat{c}_{i}\\right)\\right],\n",
+    "\\end{eqnarray}\n",
+    "where the approximation by Trotter decomposition is applied, as certain hopping terms are non-commutative.\n",
+    "The implementation of the mixed unitary indicated on the right-hand side on quantum circuits is given in Refs. [1, 2, 4, 5]."
```
Links for all references

yoshioka1128 replied:
Add links for all references.
https://github.com/tech-sketch/openqaoa/blob/yoshioka1128/dev_clean_fqaoa/examples/16_FQAOA_example.ipynb?short_path=2fe971d#L105-L106

KilianPoirier commented on examples/16_FQAOA_examples.ipynb:
Great job on the tutorial for FQAOA, this is very thorough and will be helpful for anyone trying to understand the algorithm and use it using OpenQAOA!

It would be nice to add the links directly to all references in the main text.

yoshioka1128 replied:
Thank you for your positive feedback! I’m glad to hear that you found the tutorial thorough and helpful. I appreciate your suggestion about adding direct links to all references in the main text—I'll make sure to include those in the next update.

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
```diff
@@ -0,0 +1,345 @@
+   "source": [
+    "# 17 - FQAOA circuit with advanced circuit parameterizations\n",
+    "\n",
+    "This notebook describes how to apply the annealing and Fourier parameter classes included in OpenQAOA to FAOA frame work."
```
Typo FQAOA

yoshioka1128 replied:
Fix typo FQAOA.
https://github.com/tech-sketch/openqaoa/blob/e45b04d57f3e3ca85f2b917652853984ec6ee1b1/examples/17_FQAOA_advanced_parameterization.ipynb#L10

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
```diff
@@ -0,0 +1,345 @@
+   "source": [
+    "fig = plt.figure(figsize=(10,7))\n",
+    "label_list = [f'Fourier params (p={p_fourier}, q={q}) optimization', \n",
+    "              f'Stadard params (p={p_list[0]}) optimization',]\n",
```
Typo Standard

yoshioka1128 replied:
The connection with the first half led to the decision to use fourier params and annealing params.
https://github.com/tech-sketch/openqaoa/blob/e45b04d57f3e3ca85f2b917652853984ec6ee1b1/examples/17_FQAOA_advanced_parameterization.ipynb#L321-L322

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
Same comment here, great job on this second notebook! 

Can you also add direct links to the reference here as well?

yoshioka1128 replied:
Thank you for the feedback! I’ll go ahead and add the direct links to the references in this notebook as well.

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
```diff
@@ -0,0 +1,345 @@
+    "\n",
+    "To appreciate the benefits of the Fourier parametrisation, let's compare the case $p=1$ using `StandardParams` with the case $q = 1, p=2$ using `FourierParams`. Here, we are optimising over the same total number of parameters, however the `FourierParams` ought to be capturing features of a more expressive circuit. \n",
+    "\n",
+    "Details of the Fourier parametrization are given in Ref [3] and in the Notebook `05 - QAOA circuit with advanced circuit parameterizations`."
```
Add link to [3] and Github link to the other notebook: https://github.com/entropicalabs/openqaoa/blob/dev/examples/05_advanced_parameterization.ipynb

yoshioka1128 replied:
Add links to [3] and notebook.
https://github.com/tech-sketch/openqaoa/blob/e45b04d57f3e3ca85f2b917652853984ec6ee1b1/examples/17_FQAOA_advanced_parameterization.ipynb#L245

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
```diff
@@ -0,0 +1,345 @@
+    "plt.xlabel('Number of function evaluations')\n",
+    "plt.ylabel('cost')\n",
+    "plt.legend()\n",
+    "plt.title('Comparison of FQAOA performance between Fourier and Standard Parameterizations')\n",
```
Typo Parametrizations

yoshioka1128 replied:
Fix typo Parametrizations.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/examples/17_FQAOA_advanced_parameterization.ipynb#L333

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py:
```diff
@@ -0,0 +1,402 @@
+    if n_fermions > n_qubits:
+        raise ValueError(f"n_fermions ({n_fermions}) cannot be greater than n_qubits ({n_qubits}).")
+
+    if lattice not in 'cyclic':
```
Compare to a list of value: `["cyclic"]` otherwise any sub-string is accepted, e.g. `lattice="cy"` doesn't return an error here.

yoshioka1128 replied:
Thank you for pointing that out! I have updated the code to compare against a list of values like ["cyclic"] to ensure that only exact matches are accepted. I have also added a test in test_fqaoa.py to confirm that inputs like "cy" now correctly return an error.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py#L168-L169

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py:
```diff
@@ -0,0 +1,402 @@
+        The number of fermions, which determines how many rows of the 
+        orbital matrix are considered in the computation.
+    lattice : str
+        The type of lattice configuration. Currently, only 'cyclic' lattice configurations 
+        are supported.
```
Typo: is supported

Is it by design that only `"cyclic"` is supported in `get_analytical_fermi_orbitals` but both `"cyclic"` and `"chain"` are supported in `get_fermi_orbitals`?

yoshioka1128 replied:
Fix typo is supported.

Yes, that's correct. The get_fermi_orbitals function determines the orbitals through diagonalization, while get_analytical_fermi_orbitals uses Fourier transformation to analytically find the orbitals specifically for "cyclic" cases. The reason get_analytical_fermi_orbitals does not support "chain" is that Fourier transformation cannot be applied in that context. However, to avoid confusion with condition-specific branches, I plan to standardize on using get_fermi_orbitals within fqaoa_workflow.py and keep get_analytical_fermi_orbitals as a supplementary function for testing and debugging purposes.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L346-L347

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py:
```diff
@@ -0,0 +1,402 @@
+    if lattice not in ALLOWED_LATTICE:
+        raise ValueError(f"In FQAOA, lattice {lattice} is not recognised. Please use {ALLOWED_LATTICE}")
+
+    if hopping == 0.0: raise ValueError("In FQAOA, hopping = 0 is not recgnized. Please use hopping != 0")
```
Typo: recognized

yoshioka1128 replied:
Fix typo recognized.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py#L248

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py:
```diff
@@ -0,0 +1,402 @@
+
+    return orbitals
+
+def generate_random_portfolio_data(
```
Is this function needed?
We already have a `PortfolioOptimization` problem class that incorporates a `PortfolioOptimization.random_instance` method. Do they have the same use?

yoshioka1128 replied:
`generate_random_portfolio_data` and `PortfolioOptimization.random_instance` serve similar purposes but are used differently:

- `generate_random_portfolio_data` generates random stock price fluctuations based on a specified number of days and calculates expected returns and covariance from those fluctuations.
- `PortfolioOptimization.random_instance` directly generates expected returns and covariance from random values.

The difference between them is not critical for testing or tutorial notebooks, so I have replaced all instances of `generate_random_portfolio_data` with `PortfolioOptimization.random_instance`.

KilianPoirier commented on src/openqaoa-core/tests/test_analytical_simulator.py:
```diff
@@ -185,6 +185,28 @@ def test_end_to_end_rqaoa(self):
+        # Define the device to be the analytical simulator
+        device = create_device(location="local", name="analytical_simulator")
+
+        # Define the FQAOA object and set its params
+        with self.assertRaises(ValueError):
+            fqaoa = FQAOA(device)
+
+        with self.assertRaises(ValueError):
+            fqaoa = FQAOA()
+            fqaoa.set_device(device)
```
Is the analytical simulator not available for `FQAOA`?

yoshioka1128 replied:
Yes, because XY_mixer_hamiltonian must be used in FQAOA.

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+ALLOWED_MIXERS = ["xy"]
+ALLOWED_LATTICE = ["cyclic", "chain"]
+
+class FQAOA(Workflow):
```
If `FQAOA` does not support the analytical simulator, we should probably note it here.

yoshioka1128 replied:
Based on your feedback, we've updated the code to note ALLOWED_LOCAL_SIMULATORS.
ALLOWED_LOCAL_SIMUALTORS = [
    "vectorized",
    "pyquil.statevector_simulator",
    'qiskit.qasm_simulator',
    'qiskit.shot_simulator',
    'qiskit.statevector_simulator',
]

We've added a NOT_ALLOWED_SIMULATORS list as:
NOT_ALLOWED_LOCAL_SIMULATORS = ["analytical_simulator"]
to clearly indicate which simulators are not allowed, and an appropriate ValueError is raised when an unsupported simulator like analytical is selected. 
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L49-L56

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+    >>> fqaoa.fermi_compile(problem, n_fermions)
+    >>> fqaoa.optimize()
+
+    Where `problem` is an instance of portfolio optimization and `n_fermions` is a constraint value.
```
Is there any algorithmic constraint on which type of problem we can use here?

yoshioka1128 replied:
Yes, there is an algorithmic constraint: the QUBO must include a Hamming weight constraint. Specifically, the problem should be an instance of openqaoa.problems.problem.QUBO with a constant Hamming weight, where n_fermions represents the constraint value.

To clarify this point, the following sentence has been amended as:
Where `problem` is an instance of `openqaoa.problems.problem.QUBO` with hamming weight constant constraint, where `n_fermions` is a constraint value.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L117-L118

yoshioka1128 replied:
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L117-L118

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+
+        self.backend_properties = FermiBackendProperties(**kwargs)
+
+        return None
+    @check_compiled
+
+    def set_circuit_properties(self, **kwargs):
```
Wrong indentation

yoshioka1128 replied:
Fix indentation.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L203-L204

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+            "In FQAOA, the compile(problem) method cannot be used; please use fermi_compile(problem, n_femions) instead."
+        )
+
+    def fermi_compile(
```
Is there a particular reason why we don't redefine the `compile` function?

yoshioka1128 replied:
No particular reason—fermi_compile was introduced during the coding process.
I'll use the compile method.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L261

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+
+        # Implementing SPAM Twirling and MITIQs error mitigation requires wrapping the backend.
+        # However, the BaseWrapper can have many more use cases.
+        if (
+            self.error_mitigation_properties.error_mitigation_technique
+            == "spam_twirling"
+        ):
+            self.backend = SPAMTwirlingWrapper(
+                backend=self.backend,
+                n_batches=self.error_mitigation_properties.n_batches,
+                calibration_data_location=self.error_mitigation_properties.calibration_data_location,
+            )
+        elif(
+            self.error_mitigation_properties.error_mitigation_technique
+            == "mitiq_zne"
+        ):
+            self.backend = ZNEWrapper(
+                backend=self.backend,
+                factory=self.error_mitigation_properties.factory,
+                scaling=self.error_mitigation_properties.scaling,
+                scale_factors=self.error_mitigation_properties.scale_factors,
+                order=self.error_mitigation_properties.order,
+                steps=self.error_mitigation_properties.steps
+            )
```
Just making sure, but there is no limitation to using error mitigation techniques on FQAOA?

yoshioka1128 replied:
Thank you for checking! There are no limitations to using error mitigation techniques on FQAOA. Both SPAM twirling and Mitiq ZNE can be applied without any issues.
If you have any specific concerns or scenarios in mind, please let me know, and I can provide more details!

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -0,0 +1,917 @@
+
+    @property
+    def _decomposition_standard(self) -> List[Tuple]:
+        givens_rotation = []
+        givens_rotation.append((RZ, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]))
+        givens_rotation.append((RZ, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]))
+        givens_rotation.append((RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]))
+        givens_rotation.append((X,  [self.qubit_1]))
+        givens_rotation.append((CX, [self.qubit_1, self.qubit_2]))
+        givens_rotation.append((RY, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, self.angle)]))
+        givens_rotation.append((RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, self.angle)]))
+        givens_rotation.append((CX, [self.qubit_1, self.qubit_2]))
+        givens_rotation.append((RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]))
+        givens_rotation.append((X,  [self.qubit_1]))
+        givens_rotation.append((RZ, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, -np.pi / 2)]))
+        givens_rotation.append((RZ, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, -np.pi / 2)]))
```
Since we specify everything by hand, let's write it as a list, directly like the SWAPGateMap for example:
```python
return [
  (CX, [self.qubit_1, self.qubit_2]),
  (CX, [self.qubit_2, self.qubit_1]),
  (CX, [self.qubit_1, self.qubit_2]),
]
```

I think the following would be fine:
```python
return [
  (RZ, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]),
  (RZ, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]),
  (RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]),
  (X,  [self.qubit_1]),
  (CX, [self.qubit_1, self.qubit_2]),
  (RY, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, self.angle)]),
  (RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, self.angle)]),
  (CX, [self.qubit_1, self.qubit_2]),
  (RY, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, np.pi / 2)]),
  (X,  [self.qubit_1]),
  (RZ, [self.qubit_2, RotationAngle(lambda x: x, self.gate_label, -np.pi / 2)]),
  (RZ, [self.qubit_1, RotationAngle(lambda x: x, self.gate_label, -np.pi / 2)])
]
```

yoshioka1128 replied:
I fixed it that way.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py#L708-L721

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+
+    def test_fermi_orbitals_equivalence_to_statevector(self):
+        """
+        Test that get_analytical_fermi_oribalts and get_fermi_orbitals are equivalent by outputting statevector.
```
Typo: get_analytical_fermi_orbitals

yoshioka1128 replied:
Fix the typo.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L41

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+                (5, 3, "cyclic", 1.0),
+                (5, 2, "cyclic", 1.0)
+        ]:
+            try:
+                orbitals1 = get_analytical_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+            except ValueError as e:
+                self.fail(f"Unexpected exception raised: {e}")
```
Let's rename `orbitals1` to `analytical_orbitals`.
Has this test failed for unexpected reason before? If not, let's write
```python
analytical_orbitals = get_analytical_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
```
only. In that case, if an exception is raised when calling `get_analytical_fermi_orbitals`, it's directly raised in pytest.

yoshioka1128 replied:
Thank you for the suggestion! I have renamed orbitals1(orbitals2) to `analytical_fermi_orbitals` (`fermi_orbitals`) and removed the try-except block as recommended. 
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L51-L52

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            except ValueError as e:
+                self.fail(f"Unexpected exception raised: {e}")
+
+            try:
+                orbitals2 = get_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+            except ValueError as e:
+                self.fail(f"Unexpected exception raised: {e}")
```
Same comment here, we can use `orbitals` in that case.

yoshioka1128 replied:
Thank you for the suggestion! I have renamed orbitals1(orbitals2) to analytical_fermi_orbitals (fermi_orbitals) and removed the try-except block as recommended.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L51-L52

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            except ValueError as e:
+                self.fail(f"Unexpected exception raised: {e}")
+
+            statevector = []
+            for orbitals in [orbitals1, orbitals2]:
+                statevector.append(get_statevector(orbitals))
```
We can simplify this to `statevector = [analytical_orbitals, orbitals]`.

yoshioka1128 replied:
We use simply `statevector = [get_statevector(orbitals1), get_statevector(orbitals2)]`.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L53

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            # Check statevector[0] = e^(i*theta) * statevector[1].
+            self.assertTrue(self.is_close_statevector(statevector), "Statevectors are not equivalent up to a global phase.")
+
+    def test_gicvens_rotation_angle_length(self):
```
Typo test_givens_rotation_angle_length

yoshioka1128 replied:
Fix the typo.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L57

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            lattice = params["lattice"]
+            hopping = params["hopping"]
+
+            if lattice == "cyclic" and hopping > 0.0: orbitals = get_analytical_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+            else: orbitals = get_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
```
Let's indent properly here for visibility

yoshioka1128 replied:
Use indent properly.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L71-L74

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            if lattice == "cyclic" and hopping > 0.0: orbitals = get_analytical_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+            else: orbitals = get_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+
+            try:
+                gtheta = get_givens_rotation_angle(orbitals)
+            except ValueError as e:
+                self.fail(f"Unexpected exception raised: {e}")
```
Same comment here, if there is no reason for this code to fail, then the error should be returned directly in pytest. We can simply write:
```python
gtheta = get_givens_rotation_angle(orbitals)
```

yoshioka1128 replied:
Removed the try-except block as recommended.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L76

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            lattice = params["lattice"]
+            hopping = params["hopping"]
+
+            if lattice == "cyclic" and hopping > 0.0: orbitals0 = get_analytical_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
+            else: orbitals0 = get_fermi_orbitals(n_qubits, n_fermions, lattice, hopping)
```
Same here, indent

yoshioka1128 replied:
Unified with get_fermi_orbitals and replaced with the following line.

yoshioka1128 replied:
Use indent properly.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L99-L102

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+            statevector = []
+            for i, orbitals in enumerate([orbitals0, matrix]):
+
+                try:
+                    statevector.append(get_statevector(orbitals))
+                except ValueError as e:
+                    self.fail(f"Unexpected exception raised: {e}")
```
Same thing here, no need to use `try`

yoshioka1128 replied:
Remove try and rewrite as:
`statevector = [get_statevector(orbitals0), get_statevector(matrix)]`
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L122

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+        self.assertFalse(np.array_equal(hist_exp1, hist_exp2))
+
+    @staticmethod
+    def is_close_statevector(statevector) -> bool:
+        """
+        Checks if statevector[0] can be expressed as e^(i*theta) * statevector[1].
+        """
+
+        # Threshold for considering a value to be zero
+        tolerance = 1e-10
+
+        # Check if statevector[0] is approximately zero where statevector[1] is approximately zero
+        zero_mask_1 = np.isclose(statevector[1], 0, atol=tolerance)
+        zero_mask_0 = np.isclose(statevector[0], 0, atol=tolerance)
+
+        if np.all(zero_mask_1 == zero_mask_0):
+            # Create a mask to avoid division by zero
+            non_zero_mask = ~np.isclose(statevector[1], 0, atol=tolerance)
+
+            # Compute the ratio with the mask applied
+            ratio = statevector[0][non_zero_mask] / statevector[1][non_zero_mask]
+
+            # Verify if all ratios have the same phase angle
+            theta_calculated = np.angle(ratio)
+            theta_adjusted = (theta_calculated + np.pi) % (2 * np.pi) - np.pi
+
+            # Check if the phase difference is consistent
+            consistent_phase = np.allclose(theta_adjusted, theta_adjusted[0])
+
+            if consistent_phase:
+                theta = theta_adjusted[0]
+                print(f"statevector[0] can be expressed as e^(i*theta) * statevector[1] with theta={theta}.")
+                return True
+            print("statevector[0] cannot be expressed as e^(i*theta) * statevector[1].")
+            print('theta_adjusted', theta_adjusted)
+            return False
+        print("It cannot be confirmed that statevector[0] is approximately zero when statevector[1] is approximately zero.")
+        for i in range(len(statevector[0])):
+            print(statevector[0][i], statevector[1][i])
+        return False
```
Should we change the function signature to:
```python
def is_close_statevector(statevector1, statevector2):
  ...
```

Also should we remove the print statements or do they serve a specific purpose. I suggest we make the error more verbose rather than having print statements in this sub-function.

yoshioka1128 replied:
Change the function signature to `def is_close_statevector(statevector1, statevector2)`.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_fqaoa.py#L169

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -0,0 +1,253 @@
+        and all other elements as 0.
+        """
+
+        np.set_printoptions(precision=3, suppress=True, linewidth=100)
+        print('fermi orbitals diagonalized by givens rotations')
```
Same comment here, do we need a print statement in this function or should we make the related error more verbose?

yoshioka1128 replied:
Remove the line: `print('fermi orbitals diagonalized by givens rotations')`

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+        self.assertEqual(fqaoa.circuit_properties.annealing_time, 0.7 * 2)
+        self.assertEqual(fqaoa.circuit_properties.linear_ramp_time, 0.7 * 2)
+
+    def test_set_circuit_properties_qaoa_descriptor_mixer_x(self):
+        """
+        Checks if the X mixer created by the X_mixer_hamiltonian method
+        and the automated methods in workflows do the same thing.
+
+        For each qubit, there should be 1 RXGateMap per layer of p.
+        """
+
+        nodes = 6
+        edge_probability = 0.6
+        g = nw.generators.fast_gnp_random_graph(n=nodes, p=edge_probability)
+        problem = MinimumVertexCover(g, field=1.0, penalty=10)
+
+        fqaoa = FQAOA()
+        self.assertRaises(
+            ValueError, lambda: fqaoa.set_circuit_properties(param_type="wrong name")
+        )
```
Is this test doing what the docstring describes?

yoshioka1128 replied:
Thank you for catching that! I've updated the docstring to accurately reflect the test's purpose, which is to check for a ValueError when an incorrect mixer_hamiltonian is provided. 
I also revised the code to align with this updated description.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_workflows.py#L1836-L1849

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+
+    def test_set_circuit_properties_qaoa_descriptor_mixer_x(self):
+        """
+        Checks if the X mixer created by the X_mixer_hamiltonian method
+        and the automated methods in workflows do the same thing.
+
+        For each qubit, there should be 1 RXGateMap per layer of p.
```
Also X mixer shouldn't be allowed in combination with `FQAOA`, right?

yoshioka1128 replied:
Yes, you are right.

Fix docstring and code.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_workflows.py#L1837-L1839

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+
+    def test_set_circuit_properties_variate_params(self):
+        """
+        Ensure that the Varitional Parameter Object created based on the input string , param_type, is correct.
```
Typo: Variational

yoshioka1128 replied:
Fix typo Variational.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_workflows.py#L1893

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+        Updating all attributes at the same time.
+        """
+
+        #         default_pairings = {'param_type': 'standard',
+        #                             'init_type': 'ramp',
+        #                             'qubit_register': [],
+        #                             'p': 1,
+        #                             'q': None,
+        #                             'annealing_time': 0.7,
+        #                             'linear_ramp_time': 0.7,
+        #                             'variational_params_dict': {},
+        #                             'mixer_hamiltonian': 'x',
+        #                             'mixer_qubit_connectivity': None,
+        #                             'mixer_coeffs': None,
+        #                             'seed': None}
+
+        fqaoa = FQAOA()
+
+        # TODO: Some weird error related to the initialisation of QAOA here
+        #         for each_key, each_value in default_pairings.items():
+        #             print(each_key, getattr(fqaoa.circuit_properties, each_key), each_value)
+        #             self.assertEqual(getattr(fqaoa.circuit_properties, each_key), each_value)
```
Should we remove comments?

yoshioka1128 replied:
Remove the comments.

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+
+    def test_set_circuit_properties_change(self):
+        """
+        Ensure that once a property has beefn changed via set_circuit_properties.
```
Typoe: been

yoshioka1128 replied:
Fix typo been.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_workflows.py#L1977

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+
+        self.assertRaises(AttributeError, lambda: fqaoa.backend.n_shots)
+        
+    def test_set_classical_optimizer_defaults(self):
+        pass
+
+    def test_set_classical_optimizer_jac_hess_casing(self):
+        pass
+
+    def test_set_classical_optimizer_method_selectors(self):
+        pass
+
+    def test_set_header(self):
+        pass
+
+    def test_set_exp_tags(self):
+        pass
```
Should these tests be removed?

yoshioka1128 replied:
Remove the tests.

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+        mu, sigma, _ = generate_random_portfolio_data(num_assets = num_assets, num_days = 15, seed = 1)
+        problem = PortfolioOptimization(mu, sigma, risk_factor = None, budget = budget, penalty = None).qubo
+
+        # run qaoa with different param_type, and save the objcets in a list
```
Typo: objects

yoshioka1128 replied:
Fix typo objects.
https://github.com/tech-sketch/openqaoa/blob/774e281e70481d936ee9dd9d8765d57018d79183/src/openqaoa-core/tests/test_workflows.py#L2112

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+    def test_set_exp_tags(self):
+        pass
+
+    def test_qaoa_evaluate_circuit(self):
```
Let's rename this to `test_fqaoa_evaluate_circuit`

yoshioka1128 replied:
Rename `test_fqaoa_evaluate_circuit`.
https://github.com/tech-sketch/openqaoa/blob/c731d23291166c03d7aca46a77a8e60f00d8bc3f/src/openqaoa-core/tests/test_workflows.py#L2103

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+                ):
+                fqaoa.evaluate_circuit()
+
+    def test_qaoa_evaluate_circuit_shot(self):
```
Let's rename this to `test_fqaoa_evaluate_circuit_shot`

yoshioka1128 replied:
Rename `test_fqaoa_evaluate_circuit_shot`.
https://github.com/tech-sketch/openqaoa/blob/c731d23291166c03d7aca46a77a8e60f00d8bc3f/src/openqaoa-core/tests/test_workflows.py#L2250

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+            fqaoa.set_device(device)
+            fqaoa.set_circuit_properties(p=3)
+
+            # try to evaluate the circuit before compiling
+            error = False
+            try:
+                fqaoa.evaluate_circuit()
+            except Exception:
+                error = True
+            assert (
+                error
+            ), f"param_type={param_type}. `evaluate_circuit` should raise an error if the circuit is not compiled"
```
I think this code is outdated, same in `TestingVanillaQAOA.test_evaluate_circuit_shot`, let's use the error context manager here instead:
```python
with self.assertRaises(ValueError) as cm:
  fqaoa.evaluate([1, 2, 1, 2, 1, 2])
self.assertIn(
            "Please compile the QAOA before optimizing it!",
            str(cm.exception),
        )
```
Note that `evaluate` requires to input the parameter value. We also test for the error message to make sure we raise not only the right type of error but the exact error.

Can you please also replace this in `TestingVanillaQAOA.test_evaluate_circuit_shot`?

yoshioka1128 replied:
Use error context manager here.
https://github.com/tech-sketch/openqaoa/blob/c731d23291166c03d7aca46a77a8e60f00d8bc3f/src/openqaoa-core/tests/test_workflows.py#L2266-L2271

Replace this in `TestingVanillaQAOA.test_evaluate_circuit_shot`.
https://github.com/tech-sketch/openqaoa/blob/c731d23291166c03d7aca46a77a8e60f00d8bc3f/src/openqaoa-core/tests/test_workflows.py#L1408-L1413

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+            # compile and evaluate the circuit, and check that the result is correct
+            fqaoa.fermi_compile(problem, budget)
+            result = fqaoa.evaluate_circuit([1, 2, 1, 2, 1, 2])
+            assert isinstance(
+                result["measurement_results"], dict
+            ), "When using a shot-based simulator, `evaluate_circuit` should return a dict of counts"
+            assert (
+                abs(result["cost"]) >= 0
+            ), "When using a shot-based simulator, `evaluate_circuit` should return a cost"
+            assert (
+                abs(result["uncertainty"]) > 0
+            ), "When using a shot-based simulator, `evaluate_circuit` should return an uncertainty"
```
Let's use the `unittest` functions here, `self.assertIsInstance` and `self.assertTrue`.

yoshioka1128 replied:
Use `self.assertIsInstance` and `self.assertTrue`
https://github.com/tech-sketch/openqaoa/blob/7721bd6e35a5de1048a66ddf6daf7a94e78f0c7e/src/openqaoa-core/tests/test_workflows.py#L1418-L1429

https://github.com/tech-sketch/openqaoa/blob/7721bd6e35a5de1048a66ddf6daf7a94e78f0c7e/src/openqaoa-core/tests/test_workflows.py#L2280-L2291

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+                fqaoa.backend.cvar_alpha,
+            )
+            uncertainty = np.sqrt(cost_sq - cost**2)
+            assert (
+                np.round(cost, 12) == result["cost"]
+            ), "When using a shot-based simulator, `evaluate_circuit` not returning the correct cost"
+            assert (
+                np.round(uncertainty, 12) == result["uncertainty"]
+            ), "When using a shot-based simulator, `evaluate_circuit` not returning the correct uncertainty"
```
Let's use `unittest` functions here, i.e. `self.asserTrue`

yoshioka1128 replied:
Use `self.assertTrue`.
https://github.com/tech-sketch/openqaoa/blob/7721bd6e35a5de1048a66ddf6daf7a94e78f0c7e/src/openqaoa-core/tests/test_workflows.py#L1442-L1449

https://github.com/tech-sketch/openqaoa/blob/7721bd6e35a5de1048a66ddf6daf7a94e78f0c7e/src/openqaoa-core/tests/test_workflows.py#L2304-L2311

KilianPoirier reviewed (commented):
Good job on the implementation! While the PR is pretty big I think you still managed to get something really neat and relatively easy to review.
The documentation is really helpful, you did a great lot of work on this!

There are a few typos I commented on, and other minor changes I'd like you to implement. Feel free to discuss on the comments directly.

Also note that I don't know some of the functions like Givens rotation related algorithms, I trust that you implemented them correctly.

yoshioka1128 commented:
Thank you for your review and the kind words! 
I appreciate your feedback and will go through the comments to make the suggested changes. 
I’ll also review the typos and other minor issues you pointed out.

I will pay particular attention to the functions, such as givens rotation, to make sure that they are correct.

Thanks again for your time and valuable feedback!

yoshioka1128 committed:
Fix typos and minor issues based on review comments

yoshioka1128 commented:
Thank you for the detailed review! I've pushed a new commit that addresses the typos and minor corrections as per your comments. Here's a summary of the changes:

- Fixed typos and made small corrections based on the review feedback.
- Updated variable names and improved documentation for clarity and consistency.

Please note that the error mitigation SPAM twiling is still pending and will be tackled in a separate commit.

I've also performed a final spell check using `codespell` and `aspell` to ensure there are no lingering issues. I apologize for any inconvenience caused by the previous commits. Let me know if you have any further suggestions or if anything needs additional adjustments!

yoshioka1128 committed:
Fix typos

yoshioka1128 committed:
Enable prepend_state and update tests

yoshioka1128 committed:
Reduce complexity and reorganize notebook for better structure

yoshioka1128 committed:
Refine docstrings for improved clarity and accuracy

yoshioka1128 committed:
Update notebook  for consistency with website presentation

KilianPoirier commented on examples/17_FQAOA_advanced_parameterization.ipynb:
Changes look good! We may need to change the link to reference a file on `dev` or `main` later when we merge.

yoshioka1128 replied:
Got it. I'll make sure to update the link accordingly after the merge.

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_utils.py:
```diff
@@ -16,7 +16,7 @@ def get_givens_rotation_angle(orbitals: np.array) -> List[float]:
     Parameters
     ----------
-    orbitals : np.array
+    orbitals : numpy.ndarray
```
Let's use `np.ndarray` here for consistency

yoshioka1128 replied:
This commit fixes the problem.
https://github.com/tech-sketch/openqaoa/commit/931ad4bb6b8d1a0c98b12c390ed1168727f0be9e

KilianPoirier commented on src/openqaoa-core/openqaoa/algorithms/fqaoa/fqaoa_workflow.py:
```diff
@@ -636,7 +659,7 @@ def _fermi_initial_circuit(self, orbitals: np.array, gate_applicator: object) ->
         Parameters
         ----------
-        orbitals : np.array
+        orbitals : numpy.ndarray
```
`np.ndarray`

yoshioka1128 replied:
This commit fixes the problem.
https://github.com/tech-sketch/openqaoa/commit/931ad4bb6b8d1a0c98b12c390ed1168727f0be9e

KilianPoirier commented on src/openqaoa-core/tests/test_fqaoa.py:
```diff
@@ -249,5 +166,39 @@ def is_left_aligned_diagonal_matrix(matrix: np.ndarray) -> bool:
                         return False
         return True
 
+def is_close_statevector(statevector1, statevector2) -> bool:
```
Since this is a utility, maybe it belongs in another utility file. Out of the blue I think it would make sense to put it in `openqaoa/src/openqaoa-core/openqaoa/utilities.py`

yoshioka1128 replied:
The problem is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/2ad8d6b0d10395be9ff2918ac82fb83074071a13

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -2071,23 +2035,26 @@ def test_set_backend_properties_check_backend_vectorized(self):
-        self.assertTrue(np.array_equal(fqaoa.backend.prepend_state, initial_state))
+
+        statevector = [fqaoa.backend.prepend_state, initial_state]
+        is_close_statevector(statevector[0], statevector[1])
```
Since you don't assign a variable and test right after, this can be removed

yoshioka1128 replied:
The problem is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/2ad8d6b0d10395be9ff2918ac82fb83074071a13

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -56,6 +55,7 @@
     PARAMS_CLASSES_MAPPER,
 )
 
+from test_fqaoa import is_close_statevector
```
Change the import once the function is in `utilities.py`

yoshioka1128 replied:
The problem is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/2ad8d6b0d10395be9ff2918ac82fb83074071a13

KilianPoirier commented on src/openqaoa-core/tests/test_workflows.py:
```diff
@@ -2112,54 +2078,40 @@ def test_set_backend_properties_check_backend_vectorized_w_custom(self):
-    def test_set_classical_optimizer_defaults(self):
-        pass
+        statevector = [fqaoa.backend.prepend_state, initial_state]
+        is_close_statevector(statevector[0], statevector[1])
```
Test using `self.assertTrue` ?

yoshioka1128 replied:
The problem is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/2ad8d6b0d10395be9ff2918ac82fb83074071a13

KilianPoirier reviewed (commented):
Changes look good to me! I would only like to sort out the issue with the docs and a couple of tiny comments I left.

yoshioka1128 committed:
Unified numpy.ndarray references to np.ndarray in docstrings and corrected incorrect usage of np.ndarray for np.array in the code.

yoshioka1128 committed:
refactor: remove redundant statevector variable and simplify code. prepare is_close_statevector for potential move to utilities.py

yoshioka1128 committed:
Switched from NumPy to SciPy for determinant calc

KilianPoirier commented on src/openqaoa-core/openqaoa/utilities.py:
```diff
@@ -1984,3 +1984,38 @@ def to_bin(number, n_qubits):
 
     return wavefunction
+
+def is_close_statevector(statevector1, statevector2) -> bool:
```
We need to hint the type here, both should be numpy arrays.

yoshioka1128 replied:
The issue is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/c19528bb80fd717fe789a92080c71e920d5f26be

KilianPoirier commented on src/openqaoa-core/openqaoa/utilities.py:
```diff
@@ -1984,3 +1984,38 @@ def to_bin(number, n_qubits):
+    """
+    Checks if statevector1 can be expressed as e^(i*theta) * statevector2.
+    """
+
```
Should we have a consistency check for size of the array?

yoshioka1128 replied:
The issue is fixed by following commit:
https://github.com/tech-sketch/openqaoa/commit/c19528bb80fd717fe789a92080c71e920d5f26be

KilianPoirier reviewed (commented):
Two final comments, it seems good otherwise!

yoshioka1128 committed:
Renumber notebooks and made minor corrections to notebooks.

yoshioka1128 committed:
feat: Add type hints and consistency check for statevector sizes

yoshioka1128 commented:
I want to express my sincere gratitude for your incredibly detailed review. 
Your comments have been invaluable in helping me improve the quality of this pull request. 
I'm truly grateful for your time and expertise.
If you have any further concerns or suggestions, please let me know.

KilianPoirier commented on docs/source/notebooks:
This needs to be removed

yoshioka1128 replied:
Deleted symbolic link created in the local environment.

KilianPoirier reviewed (approved):
Thank you for taking the time to address all the comments I made there. Happy to help anytime!

The PR looks good to me! Just need to remove this temporary file to make sure the docs pipeline passes.

yoshioka1128 committed:
deleted:    source/notebooks

yoshioka1128 committed:
Modify objective function to include risk factor

yoshioka1128 committed:
Update tests to account for risk factor in objective function

yoshioka1128 committed:
Adjust Jupyter notebook to reflect changes in risk factor implementation

yoshioka1128 commented:
I've added a risk factor to the objective function and updated the related tests. 
This risk factor was included in the argument but not reflected in the objective function.

Integrate risk factor into objective function and update related tests
- Modified the objective function to include risk factor in calculations.
- Updated tests to account for changes in the objective function.
- Adjusted Jupyter notebooks to reflect the new implementation of the risk factor.
